File: Merge Nodes in Between Zeros.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeNodes(ListNode* head) {
        auto ptr = head->next;
        auto new_ptr = head;
        int summ = 0;
        while (ptr) {
            while (ptr->val) {
                summ += ptr->val;
                ptr = ptr->next;
            }

            new_ptr->val = summ;
            summ = 0;
            ptr = ptr->next;
            if (ptr) {
                new_ptr = new_ptr->next;
            } else {
                new_ptr->next = nullptr;
            }
        }
        return head;
    }
};

File: Merge Nodes in Between Zeros.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeNodes(head *ListNode) *ListNode {
    ptr := head.Next
    new_ptr := head
    summ := 0
    for ptr != nil {
        for ptr.Val != 0 {
            summ += ptr.Val
            ptr = ptr.Next
        }

        new_ptr.Val = summ
        summ = 0
        ptr = ptr.Next
        if ptr == nil {
            new_ptr.Next = nil
        } else {
            new_ptr = new_ptr.Next
        }
    }
    new_ptr.Next = nil
    return head
}
